src/backend/impl_safe/database.rs

```diff
@@ -70,31 +74,36 @@ impl Snapshot {
     }
 
     pub(crate) fn put_one(&mut self, key: &[u8], value: &[u8]) {
-        let values = self.map.entry(Box::from(key)).or_insert_with(BTreeSet::new);
+        let map = Arc::make_mut(&mut self.map);
```

Comment:
Would snapshotting in SafeMode support the multithreading scenario? If so, `Arc::make_mut()` might have cloned the underlying map if there are other readers holding the read reference of `self.map`. In which case, we might want to update the `self.map` with the new one.

Reply:
That's actually exactly what Arc::make_mut already does internally.

Reply:
At least that's what my interpretation of [this](https://doc.rust-lang.org/std/sync/struct.Arc.html#method.make_mut) seems to suggest.